# **Deep Learning Tutorial - Day 1**
## **Introduction to Deep Learning**

Welcome to Day 1 of the Deep Learning journey! In this notebook, we will explore the basics of Deep Learning, understand the mathematical foundation of perceptrons, and implement a simple perceptron in Python using NumPy.

---

## **What is Deep Learning?**

Deep Learning is a subset of machine learning that focuses on neural networks with multiple layers. Unlike traditional machine learning, deep learning excels at feature extraction and representation learning from raw data, such as images, text, and audio.

### **Comparison: Deep Learning vs. Machine Learning**
| Aspect                | Machine Learning                 | Deep Learning                    |
|-----------------------|----------------------------------|----------------------------------|
| Feature Engineering   | Requires manual intervention    | Automated feature learning       |
| Data Requirements     | Works well with smaller datasets | Requires large datasets          |
| Processing Power      | Moderate requirements           | High computational requirements  |
| Performance           | Good for structured data        | Excels with unstructured data    |

---

## **Key Terminologies**
1. **Neural Network**: A computational model inspired by the human brain, consisting of neurons (nodes) connected by weights.
2. **Layers**: Neural networks are made up of input, hidden, and output layers.
3. **Weights**: Parameters that determine the strength of the connection between neurons.
4. **Activation Functions**: Functions that introduce non-linearity into the model (e.g., Sigmoid, ReLU).

---

## **Mathematics Behind Perceptrons**

A perceptron is the simplest type of neural network, consisting of:
1. **Inputs**: x1, x2, ..., xn
2. **Weights**: w1, w2, ..., wn
3. **Bias**: b (adjusts the activation threshold)
4. **Output**: The output is activated using an activation function.

### **Perceptron Formula**
The perceptron computes the following:

z = (w1 * x1) + (w2 * x2) + ... + (wn * xn) + b

The output is passed through an activation function f(z), such as Sigmoid or ReLU, to determine the final result.

---

## **Common Activation Functions**

1. **Sigmoid Function**:
   f(z) = 1 / (1 + e^(-z))

   - **Range**: (0, 1)
   - **Use case**: Outputs probabilities.

2. **ReLU (Rectified Linear Unit)**:
   f(z) = max(0, z)

   - **Range**: [0, infinity)
   - **Use case**: Avoids vanishing gradients and is computationally efficient.

---

## **Advantages and Disadvantages of Perceptrons**

### **Advantages**
1. Simple to understand and implement.
2. Forms the basis of more complex neural networks.

### **Disadvantages**
1. Cannot solve non-linear problems (e.g., XOR).
2. Requires an activation function for deeper learning tasks.


In [1]:
# Import required libraries
import numpy as np

# Define the perceptron function
def perceptron(x, w, b, activation='step'):
    """
    x: Input array (numpy array)
    w: Weights array (numpy array)
    b: Bias (float)
    activation: Activation function ('step', 'sigmoid', 'relu')
    """
    # Calculate the weighted sum
    z = np.dot(x, w) + b

    # Apply the activation function
    if activation == 'step':
        return 1 if z >= 0 else 0
    elif activation == 'sigmoid':
        return 1 / (1 + np.exp(-z))
    elif activation == 'relu':
        return max(0, z)

# Inputs and weights
x = np.array([2, 3])  # Example inputs
w = np.array([0.5, -0.5])  # Example weights
b = -1  # Example bias

# Test the perceptron with step activation
output = perceptron(x, w, b, activation='step')
print("Step Activation Output:", output)

# Test the perceptron with sigmoid activation
output = perceptron(x, w, b, activation='sigmoid')
print("Sigmoid Activation Output:", output)

# Test the perceptron with ReLU activation
output = perceptron(x, w, b, activation='relu')
print("ReLU Activation Output:", output)


Step Activation Output: 0
Sigmoid Activation Output: 0.18242552380635635
ReLU Activation Output: 0


### Step 1: Calculate
z = (0.5 * 2) + (-0.5 * 3) + (-1)

z = 1 - 1.5 - 1

z = -1.5

### Step 2: Apply Activation Functions

#### Step Activation Function:

f(z) =
- 1 if z >= 0
- 0 else

Since z = -1.5, f(z) = 0.

#### Sigmoid Activation Function:

f(z) = 1 / (1 + e^(-z))

f(z) = 1 / (1 + e^(1.5))

f(z) = 1 / (1 + 4.4817)

f(z) ≈ 1 / 5.4817 ≈ 0.1824

#### ReLU Activation Function:

f(z) = max(0, z)

f(z) = max(0, -1.5) = 0
